lib/runtime-core/src/parse.rs

```diff
@@ -206,88 +210,107 @@ pub fn read_module<
-                    mcg.feed_function_signatures(info.read().unwrap().func_assoc.clone())
+                    mcg.feed_function_signatures(info_read.func_assoc.clone())
                         .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
-                    mcg.check_precondition(&info.read().unwrap())
+                    mcg.check_precondition(&info_read)
                         .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
                 }
 
                 let fcg = mcg
-                    .next_function(Arc::clone(&info))
+                    .next_function(
+                        Arc::clone(&info),
+                        WasmSpan::new(range.start as u32, range.end as u32),
+                    )
                     .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
 
-                let info_read = info.read().unwrap();
-                let sig = info_read
-                    .signatures
-                    .get(
-                        *info
-                            .read()
-                            .unwrap()
-                            .func_assoc
-                            .get(FuncIndex::new(
-                                id as usize + info.read().unwrap().imported_functions.len(),
-                            ))
-                            .unwrap(),
-                    )
-                    .unwrap();
-                for ret in sig.returns() {
-                    fcg.feed_return(type_to_wp_type(*ret))
-                        .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
+                {
+                    let info_read = info.read().unwrap();
+                    let sig = info_read
+                        .signatures
+                        .get(
+                            *info
+                                .read()
+                                .unwrap()
+                                .func_assoc
+                                .get(FuncIndex::new(
+                                    id as usize + info_read.imported_functions.len(),
+                                ))
+                                .unwrap(),
+                        )
+                        .unwrap();
+                    for ret in sig.returns() {
+                        fcg.feed_return(type_to_wp_type(*ret))
+                            .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
+                    }
+                    for param in sig.params() {
+                        fcg.feed_param(type_to_wp_type(*param))
+                            .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
+                    }
                 }
-                for param in sig.params() {
-                    fcg.feed_param(type_to_wp_type(*param))
-                        .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
+
+                let mut operator_parser = parser.create_binary_reader();
```

Comment:
`create_binary_reader` returns a raw reader without validation: https://docs.rs/wasmparser/0.51.2/src/wasmparser/validator.rs.html#721-729 , but backends expect a pre-validated stream of events.

Reply:
Good catch, the reason is that we need the local info too and can't create a validating operator parser if we're not on `FunctionBegin` -- although, it seems like tests should definitely be failing if we disabled validation on the insides of function bodies...

If we don't have tests for that yet, we absolutely should

Reply:
I filed an issue on WasmParser to unblock this!

Reply:
Okay, this should be resolved now!